lib/vanagon/engine/amazon.rb

```diff
@@ -0,0 +1,69 @@
+require 'aws-sdk'
+require 'erb'
+require 'base64'
+require 'vanagon/engine/base'
+
+class Vanagon
+  class Engine
+    class Amazon < Base
+      attr_accessor :ami, :key_name, :userdata, :key, :key_name, :shutdown_behavior
+      attr_accessor :subnet_id, :instance_type, :vpc_id
+
+      def initialize(platform, target = nil)
+        super
+        @name = 'amazon'
+        @platform = platform
+        @ami = @platform.aws_ami
+        @target_user = @platform.target_user
+        @subnet_id = @platform.aws_subnet_id
+        @userdata = @platform.aws_user_data
+        @region = @platform.aws_region.nil? ? 'us-east-1' : @platform.region
+        @key_name = @platform.aws_key_name
+        @key = @platform.aws_key
+        @vpc_id = @platform.aws_vpc_id
+        @instance_type = @platform.aws_instance_type.nil? ? "t1.micro" : @platform.aws_instance_type
+        @required_attributes = ["ssh_port", "aws_ami", "aws_key_name"]
+        @shutdown_behavior = @platform.aws_shutdown_behavior
+
+        @ec2 = ::Aws::EC2::Client.new region: @region
+        @resource = ::Aws::EC2::Resource.new client: @ec2
+      end
+
+      def get_userdata()
+        unless @userdata.nil?
+          Base64.encode64(ERB.new(@userdata).result(binding))
+        end
+      end
+
+      def select_target
+        @instances = @resource.create_instances({
```

Comment:
If decomposed/delegated a little bit, I think it'd look a lot like this (no promises made as to the accuracy of my assumptions here; I don't know what some of these objects look like in practice):

```ruby
# Use a lazy accessor to handle instantiating the EC2
# instance and loading it into the intermediate variable
# when it's needed or returning it any time it's called
# otherwise.
def instances
  @instances ||= @resource.create_instances({
    image_id: ami,
    min_count: 1,
    max_count: 1,
    key_name: key_name,
    instance_type: instance_type,
    subnet_id: subnet_id,
    user_data: get_userdata,
    monitoring: { enabled: false }
  })
end

# You're calling `#first` on `#map` a lot and that's notably slow. Let's lazy-load
# the value that we probably care about and reference it later on.
def instance
  @instance ||= instances.first
end

def select_target
  puts "Created instance; waiting for it to start up"
  # Now with 100% less `#map`!
  @ec2.wait_until(:instance_status_ok, instance_ids: instance[:id]))
  puts "Instance created id: #{instance[:id]}"
  @target = instance[:private_ip_address]).first
rescue Aws::Waiters::Errors::WaiterFailed => error
  puts "Failed to wait for ec2 instance to start; got Amazon error #{error}"
end
```